apps/e2e/src/utils/agent.ts

```diff
@@ -0,0 +1,137 @@
+import puppeteer from 'puppeteer';
+import { PenpotApi } from './api';
+import { getFileUrl } from './get-file-url';
+
+interface Shape {
+  ':id': string;
+  ':frame-id'?: string;
+  ':parent-id'?: string;
+  ':shapes'?: string[];
+}
+
+function replaceIds(shapes: Shape[]) {
+  let id = 1;
+
+  const getId = () => {
+    return String(id++);
+  };
+
+  function replaceChildrenId(id: string, newId: string) {
+    for (const node of shapes) {
+      if (node[':parent-id'] === id) {
+        node[':parent-id'] = newId;
+      }
+
+      if (node[':frame-id'] === id) {
+        node[':frame-id'] = newId;
+      }
+
+      if (node[':shapes']) {
+        node[':shapes'] = node[':shapes']?.map((shapeId) => {
+          return shapeId === id ? newId : shapeId;
+        });
+      }
+    }
+  }
+
+  for (const node of shapes) {
+    const previousId = node[':id'] as string;
+
+    node[':id'] = getId();
+
+    replaceChildrenId(previousId, node[':id']);
+  }
+}
+
+export async function Agent() {
+  const penpotApi = await PenpotApi();
+
+  const file = await penpotApi.createFile();
+  const fileUrl = getFileUrl(file);
+
+  const browser = await puppeteer.launch({});
+  const page = await browser.newPage();
+
+  await page.setViewport({ width: 1920, height: 1080 });
+
+  page.setCookie({
+    name: 'auth-token',
+    value: penpotApi.getAuth().split('=')[1],
+    domain: 'localhost',
+    path: '/',
+    expires: (Date.now() + 3600 * 1000) / 1000,
+  });
+
+  await page.goto(fileUrl);
+  await page.waitForSelector('[data-testid="viewport"]');
+
+  page
+    .on('console', async (message) => {
+      console.log(`${message.type()} ${message.text()}`);
+    })
+    .on('pageerror', (message) => {
+      console.error(message);
+    });
+
+  const finish = async () => {
+    await penpotApi.deleteFile(file['~:id']);
+    await browser.close();
+  };
+
+  return {
+    async runCode(
+      code: string,
+      options: { screenshot?: string; autoFinish?: boolean } = {
+        screenshot: '',
+        autoFinish: true,
+      }
+    ) {
+      await page.evaluate((testingPlugin) => {
+        (globalThis as any).ɵloadPlugin({
+          pluginId: 'TEST',
+          name: 'Test',
+          code: `
+            (${testingPlugin})();
+          `,
+          icon: '',
+          description: '',
+          permissions: ['content:read', 'content:write'],
+        });
+      }, code);
+
+      await page.waitForSelector(
+        '.main_ui_workspace_right_header__saved-status',
```

Comment:
Wouldn't make sense to set a `data-xxx` for this element on core to avoid future conflicts?

Reply:
Yes, it would be better, but it's a codebase far from this. Even if I add a data attribute, it could probably be removed because it's not used in the core.